src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala

```diff
@@ -70,39 +69,69 @@ case object ThurloeDatabaseConnector extends DataAccess {
     for {
       results <- lookupWithConstraint(x => x.userId === userId)
-    } yield UserKeyValuePairs(userId, results)
+    } yield UserKeyValuePairs(userId, results map { x => x.keyValuePair })
   }
 
+  import thurloe.database.DatabaseOperation.DatabaseOperation
   /**
    * Writes the user key value pair to the database, with the exception of the 'value' which uses
    * the encrypted version instead. The IV also comes from the encryption result.
    */
-  private def databaseWrite(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes): Future[Unit] = {
-    val keyValuePairs = TableQuery[DbKeyValuePair]
+  private def databaseWrite(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes): Future[SetKeyResponse] = {
 
-    val action = keyValuePairs.insertOrUpdate(
-      None,
-      userKeyValuePair.userId,
-      userKeyValuePair.keyValuePair.key,
-      encryptedValue.base64CipherText,
-      encryptedValue.base64Iv
-    )
+    val lookupExists = keyLookup(userKeyValuePair.userId, userKeyValuePair.keyValuePair.key)
+    lookupExists flatMap { existingKvp => updateOp(existingKvp, userKeyValuePair, encryptedValue) } recoverWith {
+      case e: KeyNotFoundException => insertOp(userKeyValuePair, encryptedValue)
+      case e => Future.failed(e)
+    }
+  }
 
+  private def insertOp(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes) = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val action =
+      keyValuePairs += (
+        None,
+        userKeyValuePair.userId,
+        userKeyValuePair.keyValuePair.key,
+        encryptedValue.base64CipherText,
+        encryptedValue.base64Iv
+        )
     for {
       affectedRowsCount <- database.run(action.transactionally)
-      _ <- if (affectedRowsCount == 1) {
-        Future.successful(())
-      } else {
-        Future.failed(InvalidDatabaseStateException(
-          s"Modified $affectedRowsCount rows in database (expected to modify 1)"))
-      }
-    } yield ()
+      x <- handleDatabaseWriteResponse(affectedRowsCount, DatabaseOperation.Insert)
+    } yield x
   }
 
-  def setKeyValuePair(userKeyValuePair: UserKeyValuePair): Future[Unit] = {
+  private def updateOp(oldKeyValuePair: KeyValuePairWithId, userKeyValuePair: UserKeyValuePair, newEncryptedValue: EncryptedBytes) = {
+
+    // We've just looked up and found an entry, so this ID should never be None. However, belt and braces...
+    oldKeyValuePair.id match {
+      case None => Future.fromTry(Failure(new KeyNotFoundException(userKeyValuePair.userId, userKeyValuePair.keyValuePair.key)))
+      case Some(rowId) =>
+        // NB: Using sqlu"..." strings does clever DB magic to turn this into a proper parameterised DB command to avoid insertion attacks.
+        def sqlUpdateCommand: DBIO[Int] = sqlu"UPDATE KEY_VALUE_PAIR SET VALUE=${newEncryptedValue.base64CipherText}, IV=${newEncryptedValue.base64Iv} WHERE KVP_ID=$rowId"
```

Comment:
unfortunate that we have to resort to raw SQL, I wish I knew enough Slick to advise a specific alternative... :frowning: